meta/meta.go

```diff
@@ -1,89 +1,124 @@
+	Providers VMetadataProvidersList `json:"providers"` // list of available providers
 }
 
 type VMetadataProvider struct {
-	Name         string `json:"Name"`
-	URL          string `json:"url"`
-	Checksum     string `json:"checksum"`
-	ChecksumType string `json:"checksum_type"`
+	Name         string `json:"Name"`          // name of provider (e.g. virtualbox)
+	URL          string `json:"url"`           // url to downloadable image
+	Checksum     string `json:"checksum"`      // checksum string
+	ChecksumType string `json:"checksum_type"` // checksum type of calculated checksum string
 }
 
 // ////////////////////////////////////////////////////////////////////////////////// //
 
-func getMetaFormat(name string) string {
-	return fmtc.Sprintf("%s.json", name)
+// nameMeta returns name of the metadata file with extension
+func (m *VMetadata) nameMeta() string {
+	return fmtc.Sprintf("%s.json", m.Name)
 }
 
-//func AreEqualVersions(first string, second string) bool {
-//	return first == second
-//}
-
-//func NotEqualVersions(first string, second string) bool {
-//	return !AreEqualVersions(first, second)
-//}
-
-func (m *VMetadata) IsEmptyMeta() bool {
-	return len(m.Versions) == 0
-}
-
-func (m *VMetadata) BaseMeta() string {
+// baseMeta returns name of metadata files
+func (m *VMetadata) baseMeta() string {
 	return path.Join(m.StoragePath, m.Name)
 }
 
+// DirMeta returns directory string to metadata file
 func (m *VMetadata) DirMeta() string {
-	return path.Join(m.BaseMeta(), "metadata")
+	return path.Join(m.baseMeta(), "metadata")
 }
 
+// PathMeta returns full path to metadata file
 func (m *VMetadata) PathMeta() string {
-	return path.Join(m.DirMeta(), getMetaFormat(m.Name))
-}
-
-func (m *VMetadata) CloneMeta() *VMetadata {
-	clone := *m
-	return &clone
+	return path.Join(m.DirMeta(), m.nameMeta())
 }
 
+// URLMeta returns direct link to metadata file
 func (m *VMetadata) URLMeta() string {
 	return fmtc.Sprintf("%s/%s/%s/%s",
 		strings.Trim(m.StorageURL, "/"),
 		m.Name,
 		"metadata",
-		getMetaFormat(m.Name),
+		m.nameMeta(),
 	)
 }
 
+// HasMeta returns true if metadata file is present on disk
 func (m *VMetadata) HasMeta() bool {
 	return fsutil.IsExist(m.PathMeta())
 }
 
-func (m *VMetadata) AnyVersion(version string, f func(string, string) bool) bool {
+// IsEmptyMeta returns true if versions list is empty
+func (m *VMetadata) IsEmptyMeta() bool {
+	return m.CountVersions() == 0
+}
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+
+// Len implements interface method for Sort
+func (s VMetadataVersionsList) Len() int {
+	return len(s)
+}
+
+// Swap implements interface method for Sort
+func (s VMetadataVersionsList) Swap(i, j int) {
+	*s[i], *s[j] = *s[j], *s[i]
+}
+
+// Less implements interface method for Sort
+func (s VMetadataVersionsList) Less(i, j int) bool {
+	first, _ := version.Parse(s[i].Version)
+	second, _ := version.Parse(s[j].Version)
+
+	return first.Less(second)
```

Comment:
You can replace manual version parsing and comparison by [`VersionCompare`](https://godoc.org/pkg.re/essentialkaos/ek.v9/sortutil#VersionCompare) method from `sortutil` package.